pkg/controller/core/workload_controller.go

```diff
@@ -313,17 +313,32 @@ func isDisabledRequeuedByClusterQueueStopped(w *kueue.Workload) bool {
 }
 
 func (r *WorkloadReconciler) reconcileCheckBasedEviction(ctx context.Context, wl *kueue.Workload) (bool, error) {
-	if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadEvicted) || !workload.HasRetryOrRejectedChecks(wl) {
+	if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadEvicted) || !workload.HasRetryChecks(wl) && !workload.HasRejectedChecks(wl) {
 		return false, nil
 	}
 	log := ctrl.LoggerFrom(ctx)
 	log.V(3).Info("Workload is evicted due to admission checks")
-	workload.SetEvictedCondition(wl, kueue.WorkloadEvictedByAdmissionCheck, "At least one admission check is false")
+	if workload.HasRetryChecks(wl) {
+		workload.SetEvictedCondition(wl, kueue.WorkloadEvictedByAdmissionCheck, "At least one admission check is false")
+		err := workload.ApplyAdmissionStatus(ctx, r.client, wl, true)
+		if err == nil {
+			cqName, _ := r.queues.ClusterQueueForWorkload(wl)
+			metrics.ReportEvictedWorkloads(cqName, kueue.WorkloadEvictedByAdmissionCheck)
+		}
+		return true, client.IgnoreNotFound(err)
+	}
+	// if Workload has Rejected checks deactivate it
+	wl.Spec.Active = ptr.To(false)
+	if err := r.client.Update(ctx, wl); err != nil {
+		return false, err
+	}
+	workload.SetEvictedCondition(wl, kueue.WorkloadEvictedByDeactivation, "At least one admission check is Rejected")
```

Comment:
Could we delegate setting the Evicted condition to the workload_controller: https://github.com/kubernetes-sigs/kueue/blob/d728cb87dc3998592e1fcb27a672b15c46df8b6c/pkg/controller/core/workload_controller.go#L183-L205, and here just deactive (set active=false)? This would make the code consistent with PodsReady handling.

Reply:
Thanks for catching this, fixed!